app/jobs/google/search_keyword_job.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+module Google
+  class ClientServiceError < StandardError; end
+
+  class SearchKeywordJob < ApplicationJob
+    queue_as :default
+
+    retry_on ClientServiceError, ArgumentError
```

Comment:
What if some other exception occurs, will the job not be retried?

Reply:
I will experiment some scenarios and let you know. I know for sure the exception then bubble up to sidekiq, but I need to know exactly if sidekiq performs any retry by default or not. 
Back to you soon!

Reply:
From the logs (bellow) while testing different scenarios, we can see that it only retries for the managed cases (explicit in the `retry_on` method). 

This is interesting, and I believe we should keep the behavior as is: Some non anticipated exceptions might have a negative impact (on data or availability) and we might not want to retry. 
But retry on a `429 too many requests` is useful. 

Now I'm considering, maybe we should not retry on `ArgumentError` (ex: In case of wrong keyword_id).
What do you think @ankitkalia1195 ? :) 



### Logs for raising a standard error:
```
[ActiveJob] [Google::SearchKeywordJob] [ed9632ea-e57c-4447-b7b8-c4c99fdf461b] Performing Google::SearchKeywordJob (Job ID: ed9632ea-e57c-4447-b7b8-c4c99fdf461b) from Test(default) enqueued at  with arguments: 9
[ActiveJob] [Google::SearchKeywordJob] [ed9632ea-e57c-4447-b7b8-c4c99fdf461b] Error performing Google::SearchKeywordJob (Job ID: ed9632ea-e57c-4447-b7b8-c4c99fdf461b) from Test(default) in 0.15ms: StandardError (BAM!):
```

### Logs for raising a handled error:
```
[ActiveJob] [Google::SearchKeywordJob] [8a332423-c4ba-45c5-ac11-c25ebe5744ca] Error performing Google::SearchKeywordJob (Job ID: 8a332423-c4ba-45c5-ac11-c25ebe5744ca) from Test(default) in 12.33ms: Google::ClientServiceError (Google::ClientServiceError):
[ . . . stack trace ]
[ActiveJob] Enqueued Google::SearchKeywordJob (Job ID: 8a332423-c4ba-45c5-ac11-c25ebe5744ca) to Test(default) at 2021-07-16 10:17:32 UTC with arguments: 9
Retrying Google::SearchKeywordJob in 3 seconds, due to a Google::ClientServiceError.
``

Reply:
@malparty IMO we can skip using retry_on, and let sidekiq retry for any exception. And there are limits to the number of retires sidekiq can perform, which can be changed as well.

Reply:
Using the `retry_on` adds too many complications, when to retry, when to not, and would make the error handling more complicated as well.

Reply:
Ya, seems we can skip `retry_on`. In the sidekiq's best [handling section](https://github.com/mperham/sidekiq/wiki/Error-Handling#best-practices), its says to use sidekiq's built-in error handling and retying mechanism 
> Let Sidekiq catch errors raised by your jobs. Sidekiq's built-in retry mechanism will catch those exceptions and retry the jobs regularly. The error service will notify you of the exception. You fix the bug, deploy the fix and Sidekiq will retry your job successfully.

Reply:
That just created a new concern :D 
Now as the exception is NOT managed at rails code level, but at sidekiq level, all my tests that ensure (status: failed, html not edited, etc...) are failing due to the exception thrown while the test is running. 
I have 2 workarround, but non are `rubocop-proof`:
- wrap the test statment with begin/rescue -- not OK for Rubocop
- have 2x expect (1 expect to raise the error, 2nd expect to test the data) -- not OK for Rubocop

Do you see any clean alternative to handle this case? 
I believe it remains important to ensure
1. Status is set as failed
2. Results (html, counts, links) data is NOT being updated

Thanks for your lights @ankitkalia1195 & @junan  :)

Reply:
Regarding the testing, you mean testing the `SearchKeywordJob` directly, right?

If so then may be in the `search_job_keyword_spec.rb` we can do test for the failed case like(Pseudocode)
```Ruby
 expect { SearchKeywordJob.perform_now(id) rescue StandardError }.to change(keyword.reload, :status).to 'failed'
```

Can you try it?

And it the test is from the controller level, in that case, I think we can mock to so that it does not enqueue the job like
```Ruby
  describe 'POST #create' do
    before do
      allow(SearchKeywordJob).to receive(:perform_later).and_return(true)
    end
    
    ....
    ....
  end
```

Alternatively there is a gem([rspec-sidekiq)](https://github.com/philostler/rspec-sidekiq)) also for testing sidekiq job, but I have not used it.

Reply:
The first example is not working: The rescue behave as expected (this part is fine!) but `status` is changed in DB, not in the ruby object, so the real 'change' event is triggered when calling the `keyword.reload` method (which is NOT in the expect block).

> Though your example still indicates me that it is acceptable to rescue in a spec  -- in very specific cases like this one 👌

For perform_later:
- enqueuing is already tested in the first test & works well (exception or not) as the exception won't propagate in the test
- other tests aims to test the job logic, so I call `perform_now` which will NOT queue any job (=> it will run as a direct sync call like a basic PORO)

It is good to discover the 'rspec-sidekiq' gem, it might be handy in the future, but for this current IC, I only need the `have_enqueued_job` which is available from `include ActiveJob::TestHelper` 

### So conclusion: 
I will simply have a `begin / rescue` in these 2x specific tests. I will NOT test the retry (I rely on sidekiq tests) and we should be all good! :) 

Many thanks for this detailed insights @junan  :)